test/common/config/utility_test.cc

```diff
@@ -294,6 +294,235 @@ TEST(UtilityTest, PrepareDnsRefreshStrategy) {
+      envoy::config::core::v3::GrpcService::EnvoyGrpc config;
+      const std::string config_yaml = R"EOF(
+        cluster_name: some_xds_cluster
+    )EOF";
+      TestUtility::loadFromYaml(config_yaml, config);
+
+      config.mutable_retry_policy()->mutable_retry_back_off()->mutable_base_interval()->set_seconds(
+          test_base_interval_ms / 1000);
+      config.mutable_retry_policy()->mutable_retry_back_off()->mutable_max_interval()->set_seconds(
+          test_max_interval_ms / 1000);
+
+      JitteredExponentialBackOffStrategyPtr strategy =
+          Utility::prepareJitteredExponentialBackOffStrategy(config, random, 500, absl::nullopt);
+
+      // provided time limit is equal to max time limit
+      EXPECT_EQ(false, dynamic_cast<JitteredExponentialBackOffStrategy*>(strategy.get())
+                           ->isOverTimeLimit(test_max_interval_ms));
+
+      // provided time limit is over max time limit
+      EXPECT_EQ(true, dynamic_cast<JitteredExponentialBackOffStrategy*>(strategy.get())
+                          ->isOverTimeLimit(test_max_interval_ms + 1));
+    }
+
+    // only set base_interval
+    {
+      // Provide config values for retry
+      envoy::config::core::v3::GrpcService::EnvoyGrpc config;
+      const std::string config_yaml = R"EOF(
+        cluster_name: some_xds_cluster
+    )EOF";
+      TestUtility::loadFromYaml(config_yaml, config);
+
+      uint64_t test_base_interval_ms = 5000;
+
+      config.mutable_retry_policy()->mutable_retry_back_off()->mutable_base_interval()->set_seconds(
+          test_base_interval_ms / 1000);
+
+      JitteredExponentialBackOffStrategyPtr strategy =
+          Utility::prepareJitteredExponentialBackOffStrategy(config, random, 500, absl::nullopt);
+
+      // max_interval should be less than or equal test_base_interval * 10
+      EXPECT_EQ(false, dynamic_cast<JitteredExponentialBackOffStrategy*>(strategy.get())
+                           ->isOverTimeLimit(test_base_interval_ms * 10));
+      EXPECT_EQ(true, dynamic_cast<JitteredExponentialBackOffStrategy*>(strategy.get())
+                          ->isOverTimeLimit(test_base_interval_ms * 10 + 1));
+    }
+
+    // set max_interval < base_interval
+    {
+      uint64_t test_base_interval_ms = 10000;
+      uint64_t test_max_interval_ms = 5000;
+
+      // Provide config values for retry
+      envoy::config::core::v3::GrpcService::EnvoyGrpc config;
+      const std::string config_yaml = R"EOF(
+        cluster_name: some_xds_cluster
+    )EOF";
+      TestUtility::loadFromYaml(config_yaml, config);
+
+      config.mutable_retry_policy()->mutable_retry_back_off()->mutable_base_interval()->set_seconds(
+          test_base_interval_ms);
+      config.mutable_retry_policy()->mutable_retry_back_off()->mutable_max_interval()->set_seconds(
+          test_max_interval_ms);
+
+      EXPECT_ANY_THROW(
+          Utility::prepareJitteredExponentialBackOffStrategy(config, random, 500, absl::nullopt));
```

Comment:
Is there any test that actually test the retry behavior, rather than just that we build the correct configuration for the retry behavior?

Reply:
Well, the basic jitter backoff tests make sense as unit tests, but what I'm talking about here is some test that shows that we can both configure and observe this configuration take effect in a gRPC retry. Looking at this PR and `grpc_stream_test.cc`, I don't see any new tests or assertions covering this. FWIW you would need to use fake/mock time here to avoid flakes.

Reply:
@htuch , added unit tests in `grpc_stream_test.cc` to test gRPC retry with custom retry values.